services/QuillLMS/engines/evidence/spec/controllers/evidence/feedback_controller_spec.rb

```diff
@@ -23,6 +24,287 @@ module Evidence
+        end
+      end
+
+      context 'prefilter test' do
+        let!(:profanity_rule) do
+          create(:evidence_rule, rule_type: 'prefilter', uid: 'fdee458a-f017-4f9a-a7d4-a72d1143abeb')
+        end
+
+        before do
+          stub_const("Evidence::Check::ALL_CHECKS", [Check::Prefilter])
+        end
+
+        context 'profanity violation' do
+          it 'should return correct rule uid' do
+            post :create, params: {
+              entry: "nero was an ahole",
+              prompt_id: prompt.id,
+              session_id: 1,
+              previous_feedback: ([])
+            }, :as => :json
+            expect(response.status).to eq 200
+            parsed_response = JSON.parse(response.body)
+            expect(parsed_response['optimal']).to be false
+            expect(parsed_response['rule_uid']).to eq profanity_rule.uid
+          end
+        end
+
+        context 'no violation' do
+          it 'should return empty response' do
+            post :create, params: {
+              entry: "the quick brown fox",
+              prompt_id: prompt.id,
+              session_id: 1,
+              previous_feedback: ([])
+            }, :as => :json
+
+            expect(response.status).to eq 200
+            parsed_response = JSON.parse(response.body)
+            expect(parsed_response).to eq({})
+          end
+        end
+      end
+
+      context "grammar rule test" do
+        let(:example_error) { 'example_error' }
+        let(:example_rule_uid) { 123 }
+        let(:incoming_payload) do
+          {
+            'entry' => 'eat junk food at home.',
+            'prompt_id' => prompt.id
+          }
+        end
+
+        let(:client_response) do
+          {
+            'gapi_error' => example_error,
+            'highlight' => [{
+              'type': 'response',
+              'text': 'someText',
+              'character': 0
+            }]
+          }
+        end
+
+        let!(:grammar_rule) do
+          create(:evidence_rule, uid: example_rule_uid, optimal: false, concept_uid: 'xyz')
+        end
+
+        let!(:grammar_hint) { create(:evidence_hint, rule: grammar_rule) }
+
+        before do
+          allow(Grammar::FeedbackAssembler).to receive(:error_to_rule_uid).and_return(
+            { example_error => example_rule_uid }
+          )
+          allow_any_instance_of(Grammar::Client).to receive(:post).and_return(client_response)
+        end
+
+        it 'should return a valid json response' do
+          stub_const("Evidence::Check::ALL_CHECKS", [Check::Grammar])
+          post :create, params: incoming_payload, as: :json
+
+          expect(JSON.parse(response.body)).to eq({
+            'concept_uid' => 'xyz',
+            'feedback' => nil,
+            'feedback_type' => 'grammar',
+            'optimal' => false,
+            'response_id' => '0',
+            'highlight' => [
+              { 'type' => 'response', 'text' => 'someText', 'character' => 0 }
```

Comment:
Is it worth referencing this whole thing as `client_response['highlight']` in case those values change somehow?